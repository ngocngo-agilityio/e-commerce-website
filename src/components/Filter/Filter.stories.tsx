import type { Meta, StoryObj } from '@storybook/react';

// Components
import Filter from '.';

// Mocks
import { MOCK_CATEGORIES } from '@mocks';

const meta = {
  title: 'Filter',
  component: Filter,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof Filter>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    options: MOCK_CATEGORIES,
    defaultValue: [MOCK_CATEGORIES[0].id],
    placeholder: 'Categories',
    onChange: () => {},
  },
};
