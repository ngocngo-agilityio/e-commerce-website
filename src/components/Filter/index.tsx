// Libs
import { memo } from 'react';
import isEqual from 'react-fast-compare';
import {
  Checkbox,
  CheckboxGroup,
  Input,
  InputGroup,
  InputRightElement,
  Menu,
  MenuButton,
  MenuItem,
  MenuList,
} from '@chakra-ui/react';

// Assets
import { FilterIcon } from '@assets';

// Types
import { Category } from '@types';

interface Props {
  options: Category[];
  placeholder?: string;
  defaultValue?: string[];
  onChange: (value: string[]) => void;
}

const Filter = ({
  options = [],
  placeholder = 'Filters',
  defaultValue,
  onChange,
}: Props): JSX.Element => {
  return (
    <Menu closeOnSelect={false} size="base">
      <MenuButton>
        <InputGroup w={{ base: 'full', md: '245px' }}>
          <Input
            variant="outline"
            size="base"
            borderRadius="base"
            placeholder={placeholder}
          />
          <InputRightElement h="24px" my="12px" mr="7px">
            <FilterIcon />
          </InputRightElement>
        </InputGroup>
      </MenuButton>

      <MenuList>
        <CheckboxGroup onChange={onChange} defaultValue={defaultValue}>
          {options.map((item) => {
            const { id, name } = item || {};

            return (
              <MenuItem key={id}>
                <Checkbox
                  value={id?.toString()}
                  colorScheme="orange"
                  textTransform="capitalize"
                >
                  {name}
                </Checkbox>
              </MenuItem>
            );
          })}
        </CheckboxGroup>
      </MenuList>
    </Menu>
  );
};

export default memo(Filter, isEqual);
