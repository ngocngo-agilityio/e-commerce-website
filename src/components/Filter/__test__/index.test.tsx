// Libs
import { fireEvent, render, screen } from '@testing-library/react';

// Components
import Filter from '..';

const mockProps = {
  options: [
    { id: '1', name: 'Women' },
    { id: '2', name: 'Man' },
    { id: '3', name: 'Casual' },
  ],
  onChange: jest.fn(),
};

describe('Filter component', () => {
  test('should render Filter successfully', () => {
    const { container } = render(<Filter {...mockProps} />);

    expect(container).toBeInTheDocument();
  });

  test('should match snapshot for Filter', () => {
    const { container } = render(<Filter {...mockProps} />);

    expect(container).toMatchSnapshot();
  });

  test('should call onChange when changing options', () => {
    render(<Filter {...mockProps} />);

    const filterBtn = screen.getByPlaceholderText('Filters');

    fireEvent.click(filterBtn);
    fireEvent.click(screen.getByText(mockProps.options[0].name));

    expect(mockProps.onChange).toHaveBeenCalled();
  });
});
