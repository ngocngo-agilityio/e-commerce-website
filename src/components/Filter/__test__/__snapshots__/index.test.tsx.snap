// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter component should match snapshot for Filter 1`] = `
<div>
  <button
    aria-controls="menu-list-:r1:"
    aria-expanded="false"
    aria-haspopup="menu"
    class="chakra-menu__menu-button css-1eqi7qs"
    id="menu-button-:r1:"
  >
    <span
      class="css-xl71ch"
    >
      <div
        class="chakra-input__group css-1brjfbh"
        data-group="true"
      >
        <input
          class="chakra-input css-1lqcdac"
          placeholder="Filters"
        />
        <div
          class="chakra-input__right-element css-8p9vs9"
        >
          <svg
            class="chakra-icon css-g7h3az"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 4L10 4"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M14 4L21 4"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M3 12L12 12"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M16 12L21 12"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M3 20L8 20"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M12 20L21 20"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M10 1L10 7"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M16 9L16 15"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M8 17L8 23"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </div>
    </span>
  </button>
  <div
    class="css-0"
    style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
  >
    <div
      aria-orientation="vertical"
      class="chakra-menu__menu-list css-sjt5nk"
      id="menu-list-:r1:"
      role="menu"
      style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8);"
      tabindex="-1"
    >
      <button
        class="chakra-menu__menuitem css-wg7anz"
        data-index="0"
        id="menu-list-:r1:-menuitem-:r2:"
        role="menuitem"
        tabindex="-1"
        type="button"
      >
        <label
          class="chakra-checkbox css-t4fzte"
        >
          <input
            class="chakra-checkbox__input"
            style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
            type="checkbox"
            value="1"
          />
          <span
            aria-hidden="true"
            class="chakra-checkbox__control css-dnty2r"
          />
          <span
            class="chakra-checkbox__label css-1oeb2oe"
          >
            Women
          </span>
        </label>
      </button>
      <button
        class="chakra-menu__menuitem css-wg7anz"
        data-index="1"
        id="menu-list-:r1:-menuitem-:r3:"
        role="menuitem"
        tabindex="-1"
        type="button"
      >
        <label
          class="chakra-checkbox css-t4fzte"
        >
          <input
            class="chakra-checkbox__input"
            style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
            type="checkbox"
            value="2"
          />
          <span
            aria-hidden="true"
            class="chakra-checkbox__control css-dnty2r"
          />
          <span
            class="chakra-checkbox__label css-1oeb2oe"
          >
            Men
          </span>
        </label>
      </button>
      <button
        class="chakra-menu__menuitem css-wg7anz"
        data-index="2"
        id="menu-list-:r1:-menuitem-:r4:"
        role="menuitem"
        tabindex="-1"
        type="button"
      >
        <label
          class="chakra-checkbox css-t4fzte"
        >
          <input
            class="chakra-checkbox__input"
            style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
            type="checkbox"
            value=""
          />
          <span
            aria-hidden="true"
            class="chakra-checkbox__control css-dnty2r"
          />
        </label>
      </button>
      <button
        class="chakra-menu__menuitem css-wg7anz"
        data-index="3"
        id="menu-list-:r1:-menuitem-:r5:"
        role="menuitem"
        tabindex="-1"
        type="button"
      >
        <label
          class="chakra-checkbox css-t4fzte"
        >
          <input
            class="chakra-checkbox__input"
            style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
            type="checkbox"
            value="4"
          />
          <span
            aria-hidden="true"
            class="chakra-checkbox__control css-dnty2r"
          />
          <span
            class="chakra-checkbox__label css-1oeb2oe"
          >
            Jacket
          </span>
        </label>
      </button>
      <button
        class="chakra-menu__menuitem css-wg7anz"
        data-index="4"
        id="menu-list-:r1:-menuitem-:r6:"
        role="menuitem"
        tabindex="-1"
        type="button"
      >
        <label
          class="chakra-checkbox css-t4fzte"
        >
          <input
            class="chakra-checkbox__input"
            style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
            type="checkbox"
            value="5"
          />
          <span
            aria-hidden="true"
            class="chakra-checkbox__control css-dnty2r"
          />
          <span
            class="chakra-checkbox__label css-1oeb2oe"
          >
            Casual
          </span>
        </label>
      </button>
      <button
        class="chakra-menu__menuitem css-wg7anz"
        data-index="5"
        id="menu-list-:r1:-menuitem-:r7:"
        role="menuitem"
        tabindex="-1"
        type="button"
      >
        <label
          class="chakra-checkbox css-t4fzte"
        >
          <input
            class="chakra-checkbox__input"
            style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
            type="checkbox"
            value="6"
          />
          <span
            aria-hidden="true"
            class="chakra-checkbox__control css-dnty2r"
          />
          <span
            class="chakra-checkbox__label css-1oeb2oe"
          >
            Formal
          </span>
        </label>
      </button>
    </div>
  </div>
</div>
`;
